### ISSUE TITLE
Spotify tracklists no longer reliable when the release is unavailable

### ISSUE THREAD

**Issue by ComradeMike** [kellnerd/harmony#64]
https://harmony.pulsewidth.org.uk/release?spotify=1qGSoIDWNFmfUgyognQGNH&ts=1726603970
https://harmony.pulsewidth.org.uk/release?spotify=32nryzA6XBCX9ZUspVc1yz&ts=1729013378
https://harmony.pulsewidth.org.uk/release?spotify=2jZC33oS30WSiUSUGdd8OI&gtin=884977876475&deezer=&ts=1726774620

Unlike with other streaming services, album pages aren't deleted from Spotify when the release is removed, so it's possible to use these pages to import releases into MusicBrainz. However, as of at least a month ago (maybe even earlier), the tracklist details have become unreliable. Spotify substitutes tracks from unavailable releases for ones that are available and this seems to be happening with the API as well. This is resulting in Frankenstein releases like the ones linked above, where the track numbering is all wrong, it's split into multiple mediums and the ISRCs and track durations are questionable too. https://atisket.pulsewidth.org.uk/cached/spotify/albums_1qGSoIDWNFmfUgyognQGNH.json has the correct track details but no ISRCs, so I assume the problem is occurring when looking up additional details.

**Comment by kellnerd** [comment]
Thank you for the detailed report. A look at the cached API responses (in Harmony dev mode) revealed that only the `/v1/tracks` endpoint is affected by this bug/"feature".
Luckily it does not affect the `/v1/albums/*/tracks` endpoint, so we can still load the accurate, full tracklist of releases with more than 50 tracks.

That means the issue can be mitigated by just not using the `/v1/tracks` endpoint for unavailable releases, we basically only lose the ISRCs.
I've already implemented that in https://github.com/kellnerd/harmony/commit/72b9087b73067bf85978a78f6ae47404c773c36e, but I don't want to release that shortly before going to bed in case I made an oversight.

**Comment by kellnerd** [comment]
Thinking about this more, we should try to find a better solution.
In the past I did import unavailable releases, together with their correct ISRCs, which probably worked because Spotify was unable to substitute their tracks with anything else at all.
So we should check whether that is still possible today and under which exact conditions we can trust data from the `/v1/tracks` endpoint for unavailable releases.

**Comment by ComradeMike** [comment]
I think it's something that's changed on Spotify's end with their API, I didn't have this problem either until recently.

Interestingly, tracks 51-55 of https://harmony.pulsewidth.org.uk/release?spotify=32nryzA6XBCX9ZUspVc1yz&ts=1729013378 don't appear to be affected. I'm not sure why that is.

**Comment by kellnerd** [comment]
> Interestingly, tracks 51-55 of https://harmony.pulsewidth.org.uk/release?spotify=32nryzA6XBCX9ZUspVc1yz&ts=1729013378 don't appear to be affected. I'm not sure why that is.

That is a very important observation, it helped me to realise that the issue goes even deeper than in my previous thoughts.
The initial response for this album only contains the first 50 tracks, which means that tracks 51-55 have to be loaded through the unaffected `/v1/albums/*/tracks` endpoint.

After having had a look at more examples, the initial `/v1/albums/*` response proves to be a nasty one: While track numbers and track lengths are correct here and make everything look sensible, the track titles and IDs are taken from the replacement tracks!
So Spotify already substitutes tracks at this stage and not later when unavailable track IDs are queried. When these IDs are used with the  `/v1/tracks` endpoint to obtain ISRCs, all data is for the replacement tracks (as I noted previously).

So the trick is to not trust the tracklist from the initial `/v1/albums/*` response and **always load the full, paginated tracklist** through the  `/v1/albums/*/tracks` endpoint, even though that feels redundant for the first 50 tracks.
Then we have the correct track IDs (no substitutions) and can load the full track information including ISRC without the risk of introducing bad data.

The remaining question is whether we should always do this additional request or if it the tracklist from the initial response is sufficient as long as the release is still available.

**Comment by phw** [comment]
Is `linked_from` set for the tracks that got substituted? I guess not, the last time I checked it looked like they gave up using this field at all. But if it would be set it would include the original id, which would be great

**Comment by kellnerd** [comment]
Good point, the `linked_from` property is indeed present for the tracks of the initial response in that scenario (and only then). I don't know how I overlooked this, I guess I have compared too many track IDs from different test cases.

So scratch my second attempt as well (2d48e52d33f32224fb883e6958be8a17cdfbde05), all good things come in threes.
The final attempt to prefer the `linked_from.id` also solves the problem without an additional request and is more elegant as well (eaf4a56b26afc098580b61f32906b12428429154).
